File: carrot/carrot.py
import mongo
import protocol

import time
import math

import pika
import redis

from functools import partial
import copy
import json

from bson.objectid import ObjectId
import bson
import threading
import Queue

class Worker:
    def __init__(self, config):
        self.config = config
        self.ready = False
        self.broker = pika.BlockingConnection(pika.URLParameters(config['rabbitmq']['url']))
        self.channel = self.broker.channel()
        self.redis = redis.Redis.from_url(config['redis']['url'])
        self.mongo = mongo.connect(config['mongo'])
        self.exchangeName = ''
        self.ready = True

    def register(self, jobType, workerCallback):
        jobQueueName = self._jobQueueName(jobType)
        self._initQueueListen(jobQueueName, workerCallback)

    def reportDone(self, jobInfo, data = {}):
        print(' [x] Reporting done')
        eventInfo = {
            'eventJobType': jobInfo['jobType'],
            'eventJobId': jobInfo['jobId'],
            'eventType': protocol.DONE_EVENT_NAME,
            'data': data,
        }
        self.channel.basic_publish(exchange=self.exchangeName,
            routing_key=jobInfo['eventReturnQueue'],
            body=json.dumps(eventInfo),
            properties=pika.BasicProperties(delivery_mode=2)) # persistent message

    def reportProgress(self, jobInfo, progress, data = {}):
        p = partial(self._reportProgress, jobInfo, progress, data)
        self.broker.add_callback_threadsafe(p)

    def _reportProgress(self, jobInfo, progress, data = {}):
        print(' [x] Reporting progress')
        eventInfo = {
            'eventJobType': jobInfo['jobType'],
            'eventJobId': jobInfo['jobId'],
            'eventType': protocol.PROGRESS_EVENT_NAME,
            'progress': progress,
            'data': data,
        }
        self.channel.basic_publish(exchange=self.exchangeName,
            routing_key=jobInfo['eventReturnQueue'],
            body=json.dumps(eventInfo),
            properties=pika.BasicProperties(delivery_mode=2)) # persistent message

    def saveResult(self, jobInfo, success, output, output_ttl = None, artifacts = {}):
        p = partial(self._saveResult, jobInfo, success, output, output_ttl, artifacts)
        self.broker.add_callback_threadsafe(p)

    def _saveResult(self, jobInfo, success, output, output_ttl = None, artifacts = {}):
        # record result in redis
        _id = ObjectId(jobInfo['jobId'])
        resultKey = self._resultKey(_id)
        result = {
            'success': success,
            'output': output,
        }
        if output_ttl:
            self.redis.set(resultKey, json.dumps(result), ex=output_ttl)
        else:
            self.redis.set(resultKey, json.dumps(result))

        # record job success/failure in mongo
        mongo.update_record(
            client=self.mongo,
            db_name=self.config['mongo']['jobsDbName'],
            collection_name=self.config['mongo']['jobOutcomeCollectionName'],
            query={ '_id': _id },
            operation={
                '$set': {
                    'success': success,
                    'artifacts': artifacts,
                    'updatedAt': self._now(),
                },
                '$setOnInsert': {
                    'createdAt': self._now(),
                },
            },
            upsert=True)

        # all bookkeeping is done, the worker can acknowledge the job as done now
        return True

    def close(self):
        print(' [x] cleanup...')
        self.mongo.close()
        self.broker.close()
        # the redis.Redis connection pool (self.redis) handles disconnect

    def _threaded_callback(self, workerCallback, jobInfo, queue):
        ack = workerCallback(jobInfo)
        queue.put(ack)

    def _callback(self, workerCallback, ch, method, properties, body):
        jobInfo = json.loads(body)
        # make sure event queue exists for the worker
        # it is more efficient to do this once per job than in reportProgress/reportDone
        eventReturnQueue = jobInfo['eventReturnQueue']
        self.channel.queue_declare(eventReturnQueue, durable=True)

        # Run the callback using a thread so the main thread can handle heartbeats
        queue = Queue.Queue()
        t = threading.Thread(
            target=self._threaded_callback,
            args=(workerCallback, jobInfo, queue)
        )
        t.start()

        while t.is_alive():
            # Give pika time to process other events
            self.broker.process_data_events()

            # Pause the loop, but in a safer way for pika and broker communication
            self.broker.sleep(15)
        ack = queue.get()

        if (ack):
            self.channel.basic_ack(delivery_tag=method.delivery_tag)
        self.reportDone(jobInfo)

    def _callbackKey(self, jobType, eventType):
        return "{}-{}".format(jobType, eventType)

    def _initQueueListen(self, jobQueueName, workerCallback):
        self.channel.basic_qos(prefetch_count=1) # must be be before queue_declare and basic_consume
        self.channel.queue_declare(queue=jobQueueName, durable=True)
        boundCallback = partial(self._callback, workerCallback)
        self.channel.basic_consume(queue=jobQueueName,
            auto_ack=False,
            on_message_callback=boundCallback)
        print('consume...')
        self.channel.start_consuming()

    def _jobQueueName(self, jobType):
        return '{}_{}_{}'.format(self.config['resourcePrefix'], jobType, self.config['rabbitmq']['jobQueueSuffix'])

    def _now(self):
        # compatibility with js Date.getTime()
        return float(math.trunc(time.time()*1000.0))

    def _resultKey(self, jobId):
        return "{}_{}_{}".format(self.config['resourcePrefix'], self.config['redis']['jobResultsSuffix'], jobId)

File: carrot/js/carrot.js
/* eslint-disable no-console */

'use strict';

const EventEmitter = require('events');
const amqp = require('amqplib');
const Redis = require('ioredis');

const protocol = require('./protocol');
const mongo = require('./mongo');


class Worker extends EventEmitter {
  constructor(config) {
    super();
    this.config = config;
  }

  async init() {
    this.broker = await amqp.connect(this.config.rabbitmq.url);
    this.broker.on('error', err => this.onBrokerError(err));
    this.broker.on('end', this.onBrokerEnd);
    this.channel = await this.broker.createChannel();
    this.redis = new Redis(this.config.redis.url);
  }
  onBrokerError(err) {
    console.log(' [x] broker error');
    this.emit('error', err);
  }

  onBrokerEnd() {
    console.log(' [x] broker end');
    this.emit('end');
  }

  /** Register a worker callback for a particular job type to do the work
   * the worker will typically call saveResult() to save results.
   *
   * The worker should return true  if the job should be acknowledged as completed.
   * Otherwise, the job may be repeated.
   *
   * @param {String} jobType
   * @param {CallableFunction} workerCallback
   */
  async register(jobType, workerCallback) {
    const jobQueueName = _jobQueueName(this.config, jobType);
    await _initQueueListen(this.channel, jobQueueName, async jobInfo => {
      const jobInfoJson = _messageToJson(jobInfo);
      await this.channel.assertQueue(jobInfoJson.eventReturnQueue, { durable: true });
      const ack = await workerCallback(jobInfoJson);
      if (ack) await this.channel.ack(jobInfo);
      await this.reportDone(jobInfoJson);
    });
  }

  /** Event: Report partial progress for a job and optional extra data
   *
   * @param {Object} jobInfo
   * @param {Number} progress
   * @param {Object} data
   */
  async reportProgress(jobInfo, progress, data = {}) {
    console.log(` [x] Reporting progress`);
    const eventInfo = {
      eventJobType: jobInfo.jobType,
      eventJobId: jobInfo.jobId,
      eventType: protocol.PROGRESS_EVENT_NAME,
      progress,
      data,
    };
    await this.channel.sendToQueue(jobInfo.eventReturnQueue,
      _jsonToBuffer(eventInfo), { persistent: true });
  }

  /** Event: Report a job as complete (which is different from sending results)
   * see saveResult()
   *
   * @param {*} jobInfo
   * @param {*} data
   */
  async reportDone(jobInfo, data = {}) {
    console.log(` [x] Reporting done`);
    const eventInfo = {
      eventJobType: jobInfo.jobType,
      eventJobId: jobInfo.jobId,
      eventType: protocol.DONE_EVENT_NAME,
      data,
    };
    await this.channel.sendToQueue(jobInfo.eventReturnQueue,
      _jsonToBuffer(eventInfo), { persistent: true });
  }

  /** Record the job results to redis and that it is finished in mongo
   *
   */
  async saveResult({ jobInfoJson, success, output }) {
    // record result in redis
    const resultKey = _resultKey(this.config, jobInfoJson.jobId);
    const result = { success, output };
    await this.redis.set(resultKey, JSON.stringify(result));

    // record job success/failure in mongo
    const outcomeCollection = await mongo.getCollection({
      config: this.config.mongo,
      dbName: this.config.mongo.jobsDbName,
      collectionName: this.config.mongo.jobOutcomeCollectionName });

    await outcomeCollection.updateOne({ _id: mongo.ObjectId(jobInfoJson.jobId) },
      {
        $set: {
          success,
          updatedAt: _now(),
        },
        $setOnInsert: {
          createdAt: _now(),
        },
      }, { upsert: true });

    // all bookkeeping is done, the worker can acknowledge the job as done now
    return true;
  }

  /** close the client
   *
   */
  async close() {
    console.log(' [x] cleanup...');
    if (this.redis) await this.redis.disconnect();
    if (this.channel) await this.channel.close();
    if (this.broker) await this.broker.close();
  }
}

class Producer extends EventEmitter {
  constructor(config) {
    super();
    this.config = config;
  }

  async init() {
    this.broker = await amqp.connect(this.config.rabbitmq.url);
    this.broker.on('error', err => this.onBrokerError(err));
    this.broker.on('end', this.onBrokerEnd);
    this.channel = await this.broker.createChannel();
    this.redis = new Redis(this.config.redis.url);
    this.eventReturnQueue = this.config.eventReturnQueue;
    await this.initEventListener();
  }

  onBrokerError(err) {
    console.log(' [x] broker error');
    this.emit('error', err);
  }

  onBrokerEnd() {
    console.log(' [x] broker end');
    this.emit('end');
  }

  /** Get the result of a job by ID
   *
   * @param {String} jobId
   */
  async getResult(jobId) {
    const resultKey = _resultKey(this.config, jobId);
    try {
      return JSON.parse(await this.redis.get(resultKey));
    } catch (error) {
      console.error(error);
      throw new Error(error);
    }
  }

  /** Submit a job to the worker pool
   *
   * @param {String} jobType
   * @param {Object} args
   */
  async run({ jobType, args } = {}) {
    if (!jobType) throw new Error('jobType must be defined');

    // record job in mongo
    const jobCollection = await mongo.getCollection({
      config: this.config.mongo,
      dbName: this.config.mongo.jobsDbName,
      collectionName: this.config.mongo.jobsCollectionName });

    const eventReturnQueue = this.eventReturnQueue;
    const jobInfo = { jobType, eventReturnQueue, args };
    const res = await jobCollection.insertOne(jobInfo);

    // update job in mongo based on job id
    const jobId = res._id || res.insertedId;
    const resultKey = _resultKey(this.config, jobId);
    jobInfo.jobId = jobId;
    jobInfo.resultKey = resultKey;
    await jobCollection.updateOne({ _id: jobId },
      {
        $set: {
          resultKey,
          createdAt: _now(),
          updatedAt: _now(),
        },
      }, { upsert: false });

    // send to amqp/worker
    const jobQueue = _jobQueueName(this.config, jobType);
    await this.channel.assertQueue(jobQueue, { durable: true });
    console.log(` [x] Sending: job ${jobId} of type ${jobType} to ${jobQueue}`);
    await this.channel.sendToQueue(jobQueue, _jsonToBuffer(jobInfo), { persistent: true });
    console.log(` [x] Sent: job ${jobId}`);
    return jobId;
  }

  onJobEvent(jobType, eventType, callback) {
    const key = _callbackKey(jobType, eventType);
    console.log(` [x] registering callback for ${key}`);
    this.on(_callbackKey(jobType, eventType), info => callback(_messageToJson(info)));
  }

  onJobProgressEvent(jobType, callback) {
    return this.onJobEvent(jobType, protocol.PROGRESS_EVENT_NAME, callback);
  }

  onJobDoneEvent(jobType, callback) {
    return this.onJobEvent(jobType, protocol.DONE_EVENT_NAME, callback);
  }

  /** begin listening for job events
   *
   */
  async initEventListener() {
    await _initQueueListen(this.channel, this.eventReturnQueue, async eventInfo => {
      const eventInfoJSON = _messageToJson(eventInfo);
      console.log(` [x] event: ${JSON.stringify(eventInfoJSON)}`);
      const eventType = eventInfoJSON.eventType;
      const eventJobType = eventInfoJSON.eventJobType;
      const eventKey = _callbackKey(eventJobType, eventType);
      console.log(` [x] emit: ${eventKey}`);
      await this.emit(eventKey, eventInfo);
      await this.channel.ack(eventInfo);
    });
  }

  /** close the client
   *
   */
  async close() {
    console.log(' [x] cleanup...');
    await (await mongo.closeConnection());
    if (this.channel) await this.channel.close();
    if (this.broker) await this.broker.close();
    if (this.redis) await this.redis.disconnect();
  }
}

function _now() {
  return new Date().getTime();
}

function _callbackKey(jobType, eventType) {
  return `${jobType}-${eventType}`;
}

function _resultKey(config, jobId) {
  return `${config.resourcePrefix}_${config.redis.jobResultsSuffix}_${jobId}`;
}

function _messageToJson(message) {
  return JSON.parse(message.content.toString());
}

function _jsonToBuffer(json) {
  return Buffer.from(JSON.stringify(json));
}

async function createWorker(config) {
  const worker = new Worker(config);
  await worker.init();
  return worker;
}

async function createProducer(config) {
  const producer = new Producer(config);
  await producer.init();
  return producer;
}

function _jobQueueName(config, jobType) {
  return `${config.resourcePrefix}_${jobType}_${config.rabbitmq.jobQueueSuffix}`;
}

async function _initQueueListen(channel, queueName, callback) {
  await channel.assertQueue(queueName, { durable: true });
  await channel.prefetch(1); // evenly distribute amongst workers
  // manual acknowledgment mode,
  // see https://www.rabbitmq.com/confirms.html for details
  await channel.consume(queueName, async msg => callback(msg), { noAck: false });
}

module.exports = {
  createWorker,
  createProducer,
};
